Automattic/block-experiments
# Layout Grid: Enable mobile rendering in mobile editor.

Currently, when opening the editor on mobile, the layout grid is rendered in desktop mode.
its display is controlled by the `__experimentalGetPreviewDeviceType` property which can be set via gutenberg's preview mode, or through the block settings for the layout grid block. `__experimentalGetPreviewDeviceType` always defaults to desktop, so the layout grid always attmepts to render with maximum columns in the editor. This leads to "broken" looking pages in the mobile editor.
<img src="https://user-images.githubusercontent.com/22446385/102300152-1a178b80-3fa0-11eb-9fa5-6329d47f8f42.jpeg" width="300"/>

### Context
Initial report
https://github.com/Automattic/wp-calypso/issues/42148

More discussion and investigation
https://github.com/Automattic/block-experiments/issues/152

### Code changes and proposal
#### Goals
My aim was to keep the "Preview Mode" functionality working as it is while also supporting  editing on mobile intuitively.

#### Approach
I split the device detection into the "preview mode" represented by `selectedPreviewDeviceType` and the "rendering mode" represented by `renderDeviceType`. The preview mode remains based on the `__experimentalGetPreviewDeviceType` flag from `core/block-editor` while the render mode is based solely on the width of the "visual editor".

<img width="851" alt="Screen Shot 2020-12-16 at 1 16 25 PM" src="https://user-images.githubusercontent.com/22446385/102300778-44b61400-3fa1-11eb-9771-da5fd7f01ed1.png">

When the preview mode is changed, or the window is resized, the rendering mode is recomputed. The result is quite intuitive I think!

![Dec-16-2020 12-34-49](https://user-images.githubusercontent.com/22446385/102300898-85159200-3fa1-11eb-94d6-3cf361b870da.gif)

### Test instructions

* Sync this change to block experiments to your wordpress install or sandbox
* Add a layout grid to the editor
* Play with the preview mode, it can be changed via an icon when the block is selected, via the block settings, or via a page level preview dropdown (last option isn't available on wpcom)
* Play with the screen size, the number of columns should adjust intuitively
* What edge cases or interactions have I missed?
